src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1749,9 +1759,51 @@ AcquireExecutorShardLocksForExecution(DistributedExecution *execution)
 			 * concurrently.
 			 */
 
-			LockRelationShardResources(task->relationShardList, ExclusiveLock);
+			/*
+			 * Due to PG commit 5ee190f8ec37c1bbfb3061e18304e155d600bc8e we copy the
+			 * second parameter in pre-13.
+			 */
+			requiresConsistentSnapshotRelationShardList =
+				list_concat(requiresConsistentSnapshotRelationShardList,
+#if (PG_VERSION_NUM >= PG_VERSION_12) && (PG_VERSION_NUM < PG_VERSION_13)
+
+							list_copy(task->relationShardList));
+#else
+							task->relationShardList);
+#endif
 		}
 	}
+
+	/*
+	 * Acquire the locks in a sorted way to avoid deadlocks due to lock
+	 * ordering across concurrent sessions.
+	 */
+	anchorShardIntervalList =
+		SortList(anchorShardIntervalList, CompareShardIntervalsById);
+
+	/*
+	 * If we are dealing with a partition we are also taking locks on parent table
+	 * to prevent deadlocks on concurrent operations on a partition and its parent.
+	 */
+	LockParentShardResourceIfPartition(anchorShardIntervalList, lockMode);
```

Comment:
Well, `LockParentShardResourceIfPartition` does not call `SerializeNonCommutativeWrites` but only calls `LockShardListResources`. Not sure if we want to add that as well. For now, it is OK because modifications to partitions are not allowed for rep > 1

Reply:
Would be good to add this to the comment.